src/librustdoc/html/render/mod.rs

```diff
@@ -738,41 +738,58 @@ fn render_impls(
     w.write_str(&rendered_impls.join(""));
 }
 
-fn naive_assoc_href(it: &clean::Item, link: AssocItemLink<'_>, cx: &Context<'_>) -> String {
-    use crate::formats::item_type::ItemType::*;
+/// Build a possibly empty `href` attribute (a key-value pair) for the given associated item.
+fn assoc_href_attr(it: &clean::Item, link: AssocItemLink<'_>, cx: &Context<'_>) -> String {
+    let name = it.name.unwrap();
+    let item_type = it.type_();
 
-    let name = it.name.as_ref().unwrap();
-    let ty = match it.type_() {
-        Typedef | AssocType => AssocType,
-        s => s,
-    };
+    let href = match link {
+        AssocItemLink::Anchor(Some(ref id)) => Some(format!("#{}", id)),
+        AssocItemLink::Anchor(None) => Some(format!("#{}.{}", item_type, name)),
+        AssocItemLink::GotoSource(did, provided_items) => {
+            use ItemType::*;
+
+            let (ty_provided_item, ty_required_item, kind) = match item_type {
+                Method | TyMethod => (Method, TyMethod, ty::AssocKind::Fn),
+                AssocConst | TyAssocConst => (AssocConst, TyAssocConst, ty::AssocKind::Const),
+                AssocType | TyAssocType => (AssocType, TyAssocType, ty::AssocKind::Type),
+                _ => panic!("assoc_href_attr called on non-associated item"),
+            };
 
-    let anchor = format!("#{}.{}", ty, name);
-    match link {
-        AssocItemLink::Anchor(Some(ref id)) => format!("#{}", id),
-        AssocItemLink::Anchor(None) => anchor,
-        AssocItemLink::GotoSource(did, _) => {
-            href(did.expect_def_id(), cx).map(|p| format!("{}{}", p.0, anchor)).unwrap_or(anchor)
+            // We're creating a link from an impl-item to the corresponding
+            // trait-item and need to map the anchored type accordingly.
+            let provided = provided_items.contains(&(name, kind));
+            let ty = if provided { ty_provided_item } else { ty_required_item };
+
+            match href(did.expect_def_id(), cx) {
+                Ok((url, ..)) => Some(format!("{}#{}.{}", url, ty, name)),
+                Err(HrefError::DocumentationNotBuilt) if ty == ty_required_item => None,
+                Err(_) => Some(format!("#{}.{}", ty, name)),
+            }
         }
-    }
+    };
+
+    // links without a href are valid:
+    // https://html.spec.whatwg.org/multipage/links.html#links-created-by-a-and-area-elements
```

Comment:
They may be valid, be if we generate a link and nothing happens when you click on it, everyone will think it's a bug.

Reply:
Fair point. However, I did not invent this logic. It's already on master: https://github.com/rust-lang/rust/blob/b3e46a9763d86a1681ddaaf0e60ce537302b9c92/src/librustdoc/html/render/mod.rs#L881-L882

On master, this only done for methods, not arbitrary associated items. I just extracted that code and put it in a generalized function. I don't know how often that case occurs, it only happens if “documentation has not been built yet”: https://github.com/rust-lang/rust/blob/b3e46a9763d86a1681ddaaf0e60ce537302b9c92/src/librustdoc/html/render/mod.rs#L829

Reply:
I saw that after I commented. The code comment made me think that you added it to handle a corner case. Maybe we should add a FIXME and/or replace it with a panic... Seems super weird but it's out of context for this PR.

Reply:
Sorry for the tangled git diff